### PR TITLE
Make PipelineSpec/PipelineRef in PipelineRun consistent with TaskRun

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

In TaskRun, TaskRef can be nil, which was not the case for
PipelineRun (and PipelineRef). This fixes that by making the two consistent.

Note that we could do the opposite (aka change how it works in `TaskRun`).

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go [hunk 1]
```diff
@@ -38,17 +38,17 @@ func (ps *PipelineRunSpec) Validate(ctx context.Context) *apis.FieldError {
...
-		return apis.ErrDisallowedFields("spec.pipelineRef", "spec.pipelineSpec")
+	// can't have both pipelineRef and pipelineSpec at the same time
+	if (ps.PipelineRef != nil && ps.PipelineRef.Name != "") && ps.PipelineSpec != nil {
+		return apis.ErrDisallowedFields("spec.pipelineref", "spec.pipelinespec")
 	}
 
-	// Check that one of PipelineRef and PipelineSpec is present
-	if ps.PipelineRef.Name == "" && ps.PipelineSpec == nil {
-		return apis.ErrMissingField("spec.pipelineRef.name", "spec.pipelineSpec")
+	// Check that one of PipelineRef and TaskSpec is present
```
Comment (dibyom) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go]: Shouldn't this be PipelineSpec?

File: pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go [hunk 2]
```diff
@@ -38,17 +38,17 @@ func (ps *PipelineRunSpec) Validate(ctx context.Context) *apis.FieldError {
...
+	if (ps.PipelineRef == nil || (ps.PipelineRef != nil && ps.PipelineRef.Name == "")) && ps.PipelineSpec == nil {
+		return apis.ErrMissingField("spec.pipelineref.name", "spec.pipelinespec")
 	}
 
-	// Validate PipelineSpec if it's present
+	// Validate TaskSpec if it's present
```
Comment (dibyom) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go]: PipelineSpec?

Comment (vdemeester) [reply pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go]: Oups =)

File: pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go (hunk 1 above)
Comment (vdemeester) [reply pkg/apis/pipeline/v1alpha1/pipelinerun_validation.go]: Oups =)

**Comment by bobcatfish** [comment]
I think this might solve https://github.com/tektoncd/pipeline/issues/1498 ?

**Comment by dibyom** [comment]
> I1031 13:41:09.704] ERROR: tests timed out

Looks like YAML tests are timing out?

File: third_party/VENDOR-LICENSE [hunk 3]
```diff
@@ -5402,6 +5402,33 @@ Import: github.com/tektoncd/pipeline/vendor/github.com/modern-go/reflect2
 
 
 
+===========================================================
```
Comment (afrittoli) [inline third_party/VENDOR-LICENSE]: uh? why this change?

**Comment by afrittoli** [comment]
This will close https://github.com/tektoncd/pipeline/issues/1498

File: third_party/VENDOR-LICENSE (hunk 3 above)
Comment (vdemeester) [reply third_party/VENDOR-LICENSE]: @afrittoli not sure, ran `./hack/update-codegen.sh` and it added that :sweat:

**Comment by pritidesai** [comment]
thanks for these changes
/lgtm

File: pkg/reconciler/pipelinerun/metrics.go [hunk 4]
```diff
@@ -134,9 +134,13 @@ func (r *Recorder) DurationAndCount(pr *v1alpha1.PipelineRun) error {
 		status = "failed"
 	}
 
+	pipelineName := "anonymous"
+	if pr.Spec.PipelineRef != nil && pr.Spec.PipelineRef.Name != "" {
+		pipelineName = pr.Spec.PipelineRef.Name
+	}
 	ctx, err := tag.New(
 		context.Background(),
-		tag.Insert(r.pipeline, pr.Spec.PipelineRef.Name),
```
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/metrics.go]: The failure was legit :sweat_smile: this was putting the controller in a crashloop :stuck_out_tongue_closed_eyes:
